core/src/main/java/org/apache/accumulo/core/zookeeper/ZooSession.java

```diff
@@ -290,9 +318,19 @@ public void sync(final String path, VoidCallback cb, Object ctx) {
     verifyConnected().sync(path, cb, ctx);
   }
 
+  public void addPersistentRecursiveWatchers(Set<String> paths, Watcher watcher)
+      throws KeeperException, InterruptedException {
+    for (String path : paths) {
+      verifyConnected().addWatch(path, watcher, AddWatchMode.PERSISTENT_RECURSIVE);
+      persistentWatcherPaths.put(path, watcher);
+      log.debug("Added persistent recursive watcher at {}", path);
+    }
+  }
```

Comment:
During this loop the zookeepers could potentially change multiple times and the function could run concurrently w/ reconnect. That is making it harder to reason about these changes, could sync the method to prevent it from running at the same time as reconnect.

```suggestion
  public synchronized void addPersistentRecursiveWatchers(Set<String> paths, Watcher watcher)
      throws KeeperException, InterruptedException {
    // this is not really needed if the method is synchronized, just changed it for code clarity.
    var zk = verifyConnected();
    for (String path : paths) {
      zk.addWatch(path, watcher, AddWatchMode.PERSISTENT_RECURSIVE);
      persistentWatcherPaths.put(path, watcher);
      log.debug("Added persistent recursive watcher at {}", path);
    }
  }
```

Reply:
Made this synchronized in d6513d7